packages/postcss-reduce-initial/src/script/__tests__/io.js

```diff
@@ -0,0 +1,86 @@
+import nock from 'nock';
+import { handleError, toJSONString, write, generate } from '../lib/io.mjs';
+import testData from './sampleProperties.json';
+
+describe('toJSONString', () => {
+  const rawData = {
+    foo: 'bar',
+    'baz-qux': 'quux',
+    quuz: 'corge',
+    garply: 'waldo',
+    'fred-plugh': 'xyzzy-thud',
+  };
+
+  test('should have correct data type (string)', () => {
+    expect(typeof toJSONString(rawData)).toBe('string');
+  });
+
+  test('should produce parsable JSON', () => {
+    expect(JSON.parse(toJSONString(rawData))).toEqual(rawData);
+  });
+
+  test('should have a trailing newline', () => {
+    expect(toJSONString(rawData)).toMatch(/\n$/);
+  });
+});
+
+describe('File operations', () => {
+  const data = {
+    fromInitial: { foo: 'bar', baz: 'qux' },
+    toInitial: { qux: 'baz', bar: 'foo' },
+  };
+  const paths = { fromInitial: '/foo.json', toInitial: '/bar.json' };
+
+  test.each([
+    ['fromInitial', paths.fromInitial, data.fromInitial],
+    ['toInitial', paths.toInitial, data.toInitial],
+  ])('should write JSON file based on key (%p)', (key, path, expected) => {
+    const err = expect.any(Function);
+    const fileFunc = jest.fn();
+
+    write(fileFunc, paths, data, key);
+
+    expect(fileFunc).toHaveBeenCalledWith(path, toJSONString(expected), err);
+  });
+
+  test('should handle file operation errors', () => {
+    expect(handleError).not.toThrowError();
+    expect(() => handleError(new Error('something went wrong'))).toThrowError();
+  });
+});
+
+describe('Data fetching', () => {
```

Comment:
The intention was to cover generate and assert that at the end of whatever happens inside the function - not actually caring about the specifics of the implementation - the file function is called twice to generate two files.

For this `nock` is useful in order to remove the HTTP request from the equation.

My thinking was that this test on line 56:
`test('should make it through promise chain and write 2 files', async () => {`

... together with the assertions mentioned in [this comment](https://github.com/cssnano/cssnano/pull/1160/files#r666879350) would serve as basic smoke tests for expected behaviour without being too much tried to the implementation.

As long as the general control flow with the injected file function is preserved (allowing the test to pass in a spy) - someone could later replace `node-fetch` with something else during refactoring without breaking the test.

The thinking behind `should throw on failed API call` I think was to ensure that if  someone replace `node-fetch` in the future it doesn't fail silently when the HTTP endpoint simulated by `nock` responds with an error. To help ensure that basic error handling works.

But you are right in that `validate()` should be tested individually - and with that the `should throw on failed API call` test would anyway become redundant since `validate()` would detect the missing data regardless of whether the HTTP client knows it has failed or not.

### In summary

I would leave the first test in this suite, delete the two throw tests, and add a test for `validate()`

Reply:
I had a little look at this tonight, and using `jest.mock` features with `mockReturnValue` for the one remaining test was sufficient. So I will remove `nock` altogether like you suggest :-)

Reply:
I had a little look at this tonight, and I can replace `nock` for the one remaining test with something along the lines of:

```js
// test file: io.js
import fetch from 'node-fetch';

jest.mock('node-fetch');

// Selectively bypass mocking to use a real Response obj
const { Response } = jest.requireActual('node-fetch');

// (...)

test('should make it through promise chain and write 2 files', async () => {
  // (...)
  fetch.mockReturnValue(
    Promise.resolve(new Response(JSON.stringify(testData)))
  );
  // (...)
});
```

The disadvantage of doing so is that the test now becomes `node-fetch` specific, and thus more fragile.

What do you think?

Is removing the `nock` dev-dependency more important than the test being HTTP client agnostic?

Reply:
> I would leave the first test in this suite, delete the two throw tests, and add a test for validate()

OK.

> Is removing the nock dev-dependency more important than the test being HTTP client agnostic?

I think using `nock` just once in the project will make it harder for future maintainers to understand the tests. It's especially not worth it in my view since the cssnano package has little to do with HTTP so `nock` is not going to see much use.